### PR TITLE
add view::indices, deprecate closed_ints

### LINKED ISSUE #277: `ints` and `iota` are not compatible with similar facilities in other languages.
If you call `view::ints`/`view::iota` with a single value, it assumes that this value is the starting point of an unbounded range. This is inconsistent with similar facilities in other languages.

In Python, [`range` will assume a single value is the terminus of a half-open range starting with 0](http://pythoncentral.io/pythons-range-function-explained/). This is done to facilitate looping over integer ranges:

```
for i in range(11)
```

D's [`iota` function works the same way](http://dlang.org/phobos/std_range.html#iota). A single value means a half-open range starting at 0 and ending with the given value.

I think that the Range library versions should follow existing practice. People may find it confusing to write:

```
for(auto i : view::iota(50))
```

And get an unbounded range starting at 50.

I recognize the importance of being able to create an unbounded range. But I think it would be better to do something like this: `view::iota(50, unbounded)`.

### LINKED ISSUE #470: closed_ints is broken
It seems that `closed_ints(first, last)` maps to `iota_view(first, last + 1)` which is obviously broken if `last + 1` overflows, like here:

``` c++
auto rng = view::closed_ints((std::uint64_t)0, std::numeric_limits<std::uint64_t>::max());
// works (weirdly, because of overflow):
  std::uint64_t last0 = *(ranges::end(rng) - 1);
  CHECK(last0 == std::numeric_limits<std::uint64_t>::max());
// fails (remark: unsigned overflow is defined behavior)
  CHECK(ranges::size(rng) != (unsigned long)0); // size == 0 because of overflow
// fails (as a consequence)
  std::uint64_t last1 = 2;
  for (auto i : rng) { last1 = i; }  // because size is zero, nothing happens here
  CHECK(last1 == std::numeric_limits<std::uint64_t>::max());
```

Is there a reason why closed ints maps to `iota_view` instead of closed_iota [here](https://github.com/ericniebler/range-v3/blob/master/include/range/v3/view/iota.hpp#L494)? Why isn't `closed_iota` actually used? 

Note: `closed_ints((T)0, (T)max)` actually works for most `T`s in 64bit platforms since a _fast_ integer type is chosen (I only got it to fail for 64bit integers in my platform).

### PR DESCRIPTION
**Pull request by gnzlbg**
- Deprecate `closed_ints`. 
- add `view::indices/inclusive_indices`
- [bugfix]: fix unsigned integer wrap around which caused weird behavior like a range of size of `0` that can be iterated for`ints`/`closed_ints`/`indices`/`inclusive_indices` :
- assert on signed integer overflow by requiring `to - from < std::numeric_limits<T>::max()` for`ints`/`closed_ints`/`indices`/`inclusive_indices`

Closes #470 (at least the "bug" part of it. @CaseyCarter: it might be worth it to move your ideas for infinite ranges to its own issue to pursue them further).

Closes #277 .

Supersedes #592 .

### CODE REVIEW

File: include/range/v3/view/iota.hpp [hunk 1]
```diff
@@ -524,6 +524,7 @@ namespace ranges
                 detail::take_exactly_view_<iota_view<Val>, true>>
             ints_fn::operator()(Val from, Val to) const
             {
+                RANGES_EXPECT(to >= from);
```
Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: I disagree that we should require `to >= from`, at least for unsigned types. I don't particularly care about `[0, -2)`, since we already have `view::iota(0)` to name that infinite ranges, but `[(1u << 30), 2u)` is a perfectly well-defined range of unsigned integers.

This could easily be `RANGES_EXPECT(UnsignedIntegral<Val>() || to >= from);`

File: include/range/v3/view/iota.hpp [hunk 2]
```diff
@@ -533,6 +534,8 @@ namespace ranges
                     CONCEPT_REQUIRES_(Integral<Val>())>
                 detail::take_exactly_view_<iota_view<Val>, true> operator()(Val from, Val to) const
                 {
+                    RANGES_EXPECT(to >= from);
```
Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: Ditto.

File: include/range/v3/view/iota.hpp (hunk 1 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: @CaseyCarter note that `[(1u << 30), 2u)` is currently broken in range-v3, because `iota_minus_` returns a negative integer for the distance (it performs the substraction correctly), so `take_exactly` will fail when passed a negative integer. . . 

I've made some changes to the PR to fix this, that hopefully do not break any other cases (I've added some new tests for unsigned wrap around as well). 

Please take a careful look at `iota_distance_` and let me know what you think. 

The good think from doing things this way is that we catch with an assertion all cases in which weird behavior would happen. 

The thing that worries me the most is that `view::take_exactly` takes a `difference_type`, which means that for ranges of unsigned integers, a range size of at most the largest representable value by the unsigned integers difference type can be produced. For 32 bit unsigned integers on 32 bit platforms this means that the range shall contain at most `std::numeric_limits<uint32_t>::max() / 2` elements.

File: include/range/v3/view/iota.hpp [hunk 3]
```diff
@@ -533,7 +552,7 @@ namespace ranges
                     CONCEPT_REQUIRES_(Integral<Val>())>
                 detail::take_exactly_view_<iota_view<Val>, true> operator()(Val from, Val to) const
                 {
-                    return {iota_view<Val>{from}, detail::iota_minus_(to, from) + 1};
+                  return {iota_view<Val>{from}, detail::ints_distance_(from, to) + 1};
```
Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: s/return/  return/

Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: fixed

File: test/view/iota.cpp [hunk 4]
```diff
@@ -50,6 +50,22 @@ void test_iota_minus() {
...
+
+    auto rng = view::ints(from, to);
+    CHECK(ranges::size(rng) == std::size_t(5));
+    ::check_equal(rng, {U(max - 2), U(max - 1), max, U(0), U(1)});
+
+    auto rng2 = view::closed_ints(from, to);
+    CHECK(ranges::size(rng2) == std::size_t(6));
+    ::check_equal(rng2, {U(max - 2), U(max - 1), max, U(0), U(1), U(2)});
+}
+
```
Comment (ericniebler) [inline test/view/iota.cpp]: This seems like a perversion to me. And now I feel bad because I didn't stop you before spending time on it. :-( I would prefer a narrow interface on `iota` and `ints`^D^D^D^D`indices` that requires `from < to`.

**Comment by ericniebler** [comment]
I don't feel confident making _any_ change to `ints` or `closed_ints` since the semantics of edge cases are weird. And I don't feel motivated to make any changes since we've decided elsewhere to deprecate them. My recommendation is to just leave them alone and hurry up deprecating them.

File: test/view/iota.cpp (hunk 4 above)
Comment (gnzlbg) [inline test/view/iota.cpp]: While this is indeed a perversion, I think this should work for unsigned integers, and that we should assert for signed integers to prevent undefined behavior. 

I am preparing right now a PR that deprecates this and adds `view::indices/closed_indices` and for those and unsigned integers I use this behavior as well. 

Maybe fixing these for the two-argument ints version is not worth it since those who could benefit will almost immediately get a deprecated warning, but the fix is already here so... I also don't see a reason why we shouldn't apply it :/

**Comment by gnzlbg** [comment]
So i've added `view::indices/closed_indices` and deprecated `closed_ints`.

I've preserved the "fixed" wrap around semantics and test for unsigned integers because "why not".

I've made the default `indices/closed_indices` integer type be `std::size_t` because that's the types of indices in the STL. Thoughts on this? Should I remove this behavior from `view::indices` all together?

File: include/range/v3/view/iota.hpp [hunk 5]
```diff
@@ -163,6 +163,25 @@ namespace ranges
                     :  static_cast<D>(v0 - v1);
             }
 
+            template<typename Val, CONCEPT_REQUIRES_(SignedIntegral<Val>())>
+            RANGES_CXX14_CONSTEXPR
+            iota_difference_t<Val> ints_distance_(Val from, Val to) noexcept {
+                using D = iota_difference_t<Val>;
+                RANGES_EXPECT(from <= to);
+                RANGES_EXPECT((static_cast<D>(to) - static_cast<D>(from))
+                              < std::numeric_limits<D>::max());
```
Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: What does this expectation accomplish? If `to - from == max`, it's representable. `to - from <= max` always holds.

Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: If it's an attempt to guard against `to - from` overflowing `D`, which is commendable, it's not working.

File: include/range/v3/view/iota.hpp [hunk 6]
```diff
@@ -163,6 +163,25 @@ namespace ranges
...
+                return static_cast<D>(to) - static_cast<D>(from);
+            }
+
+            template<typename Val, CONCEPT_REQUIRES_(UnsignedIntegral<Val>())>
+            RANGES_CXX14_CONSTEXPR
+            iota_difference_t<Val> ints_distance_(Val from, Val to) noexcept {
+                using D = iota_difference_t<Val>;
+                RANGES_EXPECT(static_cast<D>(to - from) < std::numeric_limits<D>::max());
```
Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: Ditto.

File: test/view/iota.cpp [hunk 7]
```diff
@@ -34,20 +35,40 @@ struct Int
...
+    CHECK(detail::iota_minus_(I(max - I(1)), I(max - I(1))) == D(0));
+    CHECK(detail::iota_minus_(I(max - I(1)), I(max)) == D(-1));
+    CHECK(detail::iota_minus_(I(max), I(max - I(1))) == D(1));
+    CHECK(detail::iota_minus_(I(max), I(max)) == D(0));
+}
+
+template <typename U>
+void test_wrap_around() {
+    using namespace ranges;
+    auto max = std::numeric_limits<U>::max();
```
Comment (CaseyCarter) [inline test/view/iota.cpp]: I would get rid of `max`, and use `U(-1)`, `U(-2)`, and `U(-3)` in place of `max`, `U(max - 1)`, and `U(max - 2)`. This function then behaves identically for both signed and unsigned types.

File: include/range/v3/view/indices.hpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+            /// Half-open range of indices: [from, to).
+            struct indices_fn
+              : iota_view<std::size_t>
+            {
+                indices_fn() = default;
+
+                template<typename Val>
+                meta::if_c<
+                    (bool)Integral<Val>(),
+                    detail::take_exactly_view_<iota_view<Val>, true>>
```
Comment (CaseyCarter) [inline include/range/v3/view/indices.hpp]: Why is this not `take_exactly_view<iota_view<Val>>`? (and on 80)

File: include/range/v3/view/indices.hpp [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        namespace view
+        {
+            /// Half-open range of indices: [from, to).
+            struct indices_fn
+              : iota_view<std::size_t>
+            {
+                indices_fn() = default;
+
+                template<typename Val>
+                meta::if_c<
```
Comment (CaseyCarter) [inline include/range/v3/view/indices.hpp]: Why is the style of constraint different for this function and the function on line 50?

File: include/range/v3/view/iota.hpp [hunk 10]
```diff
@@ -524,16 +543,17 @@ namespace ranges
...
 
             struct closed_ints_fn
             {
                 template<typename Val,
                     CONCEPT_REQUIRES_(Integral<Val>())>
+                RANGES_DEPRECATED("view::closed_ints is deprecated; use view::closed_indices instead!")
                 detail::take_exactly_view_<iota_view<Val>, true> operator()(Val from, Val to) const
                 {
-                    return {iota_view<Val>{from}, detail::iota_minus_(to, from) + 1};
+                    return {iota_view<Val>{from}, detail::ints_distance_(from, to) + 1};
```
Comment (CaseyCarter) [inline include/range/v3/view/iota.hpp]: Is this why `ints_distance_` forbids `to - from == max`? If so, that precondition should be here instead of there.

File: test/view/iota.cpp [hunk 11]
```diff
@@ -76,20 +97,20 @@ int main()
     models<concepts::BoundedView>(aux::copy(shorts));
     static_assert(Same<int, range_difference_type_t<decltype(shorts)>>(), "");
 
-    auto uints = view::closed_ints(
+    auto uints = view::inclusive_indices(
         std::numeric_limits<std::uint32_t>::min(),
```
Comment (CaseyCarter) [inline test/view/iota.cpp]: s/(u)?int([0-9]+)_t/\1int_least\2_t/g

File: test/view/iota.cpp [hunk 12]
```diff
@@ -76,20 +97,20 @@ int main()
...
-        std::numeric_limits<std::uint32_t>::max());
+        std::numeric_limits<std::uint32_t>::max() - 1);
     models<concepts::BoundedView>(aux::copy(uints));
     static_assert(Same<std::int64_t, range_difference_type_t<decltype(uints)>>(), "");
```
Comment (CaseyCarter) [inline test/view/iota.cpp]: s/int64_t/intmax_t/

File: test/view/iota.cpp [hunk 13]
```diff
@@ -76,20 +97,20 @@ int main()
...
     static_assert(Same<std::uint64_t, range_size_type_t<decltype(uints)>>(), "");
```
Comment (CaseyCarter) [inline test/view/iota.cpp]: s/uint64_t/uintmax_t/

File: include/range/v3/view/indices.hpp [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+namespace ranges
+{
+    inline namespace v3
+    {
+
+        namespace view
+        {
+            /// Half-open range of indices: [from, to).
+            struct indices_fn
+              : iota_view<std::size_t>
```
Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: If we are going to use `difference_type` for indexing in the STL2, we should use `std::ptrdiff_t` here.

File: include/range/v3/view/indices.hpp (hunk 9 above)
Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: @ericniebler ?

I just took this from `view::ints`, and since I did not know _why_ they were constrained that way instead of with `CONCEPT_REQUIRES_` I decided not to change it, but I see no reason against using `CONCEPT_REQUIRES_` here.

File: include/range/v3/view/iota.hpp (hunk 6 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: See my comment above. While this check is probably wrong, `to - from < std::numeric_limits<range_size_t>::max()` or otherwise computing the size will overflow and give `0`, so I think fixing the check here makes sense. Otherwise we end up with ranges of size `0` that are iterable ...

File: include/range/v3/view/iota.hpp (hunk 5 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: So this check uses `<` instead of `<=` to catch when the range size which is `(to - from) + 1` for closed ranges would overflow. That is, if `to - from == max` everything is still fine, but when one attempts to compute the size that will overflow for _unsigned integers_.

However, we are dealing with signed integers here which have an unsigned range size that is at least as wide, so doing that should never overflow, so this check is moot and can be removed. 

> If it's an attempt to guard against to - from overflowing D, which is commendable, it's not working.

I don't think that can ever work. If `to - from` overflows its UB, I don't see a way to detect this.

File: include/range/v3/view/iota.hpp (hunk 10 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: Yep. I remember now why I used difference type, it's because `take_exactly` takes a difference type, so that one is the one in which overflow must be prevented. 

I think I should write a comment because I did it 2 days ago and already forgot why. . . sigh.

File: test/view/iota.cpp (hunk 7 above)
Comment (gnzlbg) [inline test/view/iota.cpp]: I am only using this function below to test the unsigned types and thought `max` was more clear than `U(-1)`.

File: include/range/v3/view/indices.hpp (hunk 9 above)
Comment (ericniebler) [inline include/range/v3/view/indices.hpp]: > Why is the style of constraint different for this function and the function on line 50?
  
I dunno. :-|

File: include/range/v3/view/indices.hpp (hunk 8 above)
Comment (ericniebler) [inline include/range/v3/view/indices.hpp]: I dunno.

File: include/range/v3/view/indices.hpp (hunk 9 above)
Comment (ericniebler) [inline include/range/v3/view/indices.hpp]: Oh, I remember. I needed to fall back on `enable_if`-style constraints when I needed the ability to forward-declare a constrained function.

**Comment by ericniebler** [comment]
> I've preserved the "fixed" wrap around semantics and test for unsigned integers because "why not".

Why not: Because users are far, far, *far* more likely to accidentally get wrap-around behavior in error than to actually want unsigned wrap-around. I would much prefer to trap these bugs than to enable a use case of highly dubious value.

File: test/view/iota.cpp (hunk 7 above)
Comment (CaseyCarter) [inline test/view/iota.cpp]: The fact that `[U(-3), U(2))` is a valid range of exactly 5 elements whether or not `U` is a signed or unsigned integral type is the primary motivation for not requiring `to < from` for ranges of unsigned types.

Comment (ericniebler) [inline test/view/iota.cpp]: https://github.com/ericniebler/range-v3/pull/593#issuecomment-287879747

File: include/range/v3/view/indices.hpp (hunk 14 above)
Comment (CaseyCarter) [inline include/range/v3/view/indices.hpp]: > That said, I'm a bit uncomfortable with drifting from unsigned to signed here and there, in dribs and drabs.

I have an eventual goal in mind. I've got a rough branch that makes `ranges::size` always return a signed type, and changes all of the `size_type`s of the views to be signed types. I've got the beginnings of a semantic wrapper `nonnegative<T>` that "knows" that the stored `T` value is not negative. What I haven't done yet is wrap all of that up nicely to see how it works together and interacts with the standard library.

I think your "Dribs and drabs" argument has sold me. Changing this component to use a signed type as well is not a significant burden, and I'd much rather present a holistic design than try to explain why a single component is trying to swim against the current.

Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: I've changed this to use `std::ptrdiff_t` instead of `std::size_t` as default integer type.

File: include/range/v3/view/indices.hpp (hunk 8 above)
Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: @CaseyCarter @ericniebler because of [this](https://github.com/ericniebler/range-v3/blob/master/include/range/v3/view/take_exactly.hpp#L43), and [STL2 issue 63](https://github.com/ericniebler/stl2/issues/63).

File: include/range/v3/view/indices.hpp (hunk 9 above)
Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: I've switched this to `CONCEPT_REQUIRES_` for view::indices/closed_indices.

File: include/range/v3/view/iota.hpp (hunk 6 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: So I've removed the check in the signed version, and hopefully fixed this check.

File: test/view/iota.cpp (hunk 12 above)
Comment (gnzlbg) [inline test/view/iota.cpp]: fixed

File: test/view/iota.cpp (hunk 13 above)
Comment (gnzlbg) [inline test/view/iota.cpp]: fixed

**Comment by gnzlbg** [comment]
@ericniebler I agree that this is often a bug. The only concern I have is: "what about users that do want wrap around behavior"? While I don't think that we should provide a solution for this right now, I do think that we should at least discuss possible paths forward to supporting those, and try to make sure that those can be pursued by whoever is interested without breaking backwards compatibility (within reasonable margins, range-v3 is not the C++ standard). 

Here is my mental sketch of how somebody interested _could_ extend `view::indices` to support wrap around behavior. 

- 0. `view::indices` currently only supports native integral types (`std::is_integral<T>`), and does not allow wrap around (as of this commit). This would be the starting point after this PR.
- 1. Fix part of #591. That is, add an `IntegralNumber` concept that allows "user-defined" integer types with same complexities and semantics as `std::is_integral" native types. 
- 2. Add an user-defined `wrapping<T>` library type that models `IntegralNumber` if T models `IntegralNumber` and provides wrapping integer arithmetic independently of signed-ness.
- 3. Extend in a backwards-compatible way `view::indices`/`ints` to `IntegralNumber` types, that use whatever wrap-around behavior user-defined integral numbers provide. This will require dancing around `std::is_integral` types also modeling `IntegralNumber`, but is doable in a backwards compatible way. 

That is, those wanting wrap around behavior would just need to pass their wrapping types to `view::indices` or write `view::indices(wrapping(a), wrapping(b))`.

File: include/range/v3/view/iota.hpp [hunk 15]
```diff
@@ -163,6 +163,47 @@ namespace ranges
...
+                    "iota_difference_type must be at least as wide as the signed integer type; "
+                    "otherwise the expression below might overflow when to - from would not.");
+                return static_cast<D>(to) - static_cast<D>(from);  
+            }
+
+            template<typename Val, CONCEPT_REQUIRES_(UnsignedIntegral<Val>())>
+            RANGES_CXX14_CONSTEXPR
+            iota_difference_t<Val> ints_distance_(Val from, Val to) noexcept {
+                using D = iota_difference_t<Val>;
+                using UD = std::make_unsigned_t<D>;
```
Comment (ericniebler) [inline include/range/v3/view/iota.hpp]:
```
/Users/travis/build/ericniebler/range-v3/include/range/v3/view/iota.hpp:181:33: error: no template named 'make_unsigned_t' in namespace 'std'; did you mean 'make_unsigned'?
                using UD = std::make_unsigned_t<D>;
                           ~~~~~^~~~~~~~~~~~~~~
                                make_unsigned
```

File: include/range/v3/view/indices.hpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+                (
+                    detail::take_exactly_view_<iota_view<Val>, true>
+                        {iota_view<Val>{from}, detail::ints_distance_(from, to)}
+                )
+
+                template<typename Val, CONCEPT_REQUIRES_(Integral<Val>())>
+                auto operator()(Val to) const
+                RANGES_DECLTYPE_AUTO_RETURN_NOEXCEPT
+                (
+                    (*this)(Val(), to)
```
Comment (ericniebler) [inline include/range/v3/view/indices.hpp]: Older versions of gcc will not suffer to have `this` appear in trailing return types.

File: include/range/v3/view/indices.hpp [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            RANGES_INLINE_VARIABLE(indices_fn, indices)
+
+            /// \relates inclusive_indices_fn
+            /// \ingroup group-views
+            RANGES_INLINE_VARIABLE(inclusive_indices_fn, inclusive_indices)
+
+        }  // namespace view
+    }
+}
+
```
Comment (ericniebler) [inline include/range/v3/view/indices.hpp]: You need to do the `SATISFY_BOOST_RANGE` dance.

File: include/range/v3/view/iota.hpp (hunk 10 above)
Comment (ericniebler) [inline include/range/v3/view/iota.hpp]: If we're always doing `n+1` in `closed_ints` and `inclusive_indices`, then the need for the view goes away. Users can do `+1` themselves. What they *can't* do it get a closed range that covers the entire range of integers without fear of overflow.

If we decide that's not a valid thing to want, I suggest we deprecate `closed_ints` and `inclusive_indices`. (And I disapprove of the inconsistent naming, FWIW.) If it *is* a valid thing to want, then this PR is wrong.

File: include/range/v3/view/indices.hpp (hunk 8 above)
Comment (CaseyCarter) [inline include/range/v3/view/indices.hpp]: [Looking in range_fwd.hpp](https://github.com/CaseyCarter/range-v3/blob/master/include/range/v3/range_fwd.hpp#L658), `take_exactly_view<iota_view<Val>>` is still defined as `detail::take_exactly_view_<iota_view<Val>, is_random_access_bounded_<Rng>::value>>` and [`is_random_access_bounded_` is still `RandomAccess` and `BoundedRange`](https://github.com/CaseyCarter/range-v3/blob/master/include/range/v3/view/take_exactly.hpp#L39).

Until `take_exactly_view` changes, there's no reason to break the abstraction here.

**Comment by CaseyCarter** [comment]
> I agree that this is often a bug. The only concern I have is: "what about users that do want wrap around behavior"? 

In a future where people use unsigned integers because they *want* modulo behavior, and not because they want "nonnegative," we would re-enable the proper behavior here.

File: include/range/v3/view/indices.hpp (hunk 17 above)
Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: IIUC `indices` is not a "view", it just returns `iota | take_exactly` and `iota` already does this, so that should be enough right?

File: include/range/v3/view/iota.hpp (hunk 15 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: fixed

File: include/range/v3/view/iota.hpp (hunk 10 above)
Comment (gnzlbg) [inline include/range/v3/view/iota.hpp]: @ericniebler I've split `ints_distance_` into `ints_open_distance_` and `ints_closed_distance`. 

Its not about "jus" doing `+1`, but rather, about making sure that this `+1` doesn't overflow the signed difference_type that `view::take_exactly` takes as argument. 

So for open ranges, one checks that:
- `from <= to` (for both signed and unsigned integers),
- `Unsigned(to - from) <= Unsigned(numeric_limits<difference_type>::max())` for unsigned integers only, otherwise overflowing the signed difference type is undefined behavior.

For closed ranges one has the same checks as for open ranges, plus:
- `open_distance < numeric_limits<difference_type>::max()` (for both)
since otherwise `open_distance + difference_type(1) will overflow, introducing undefined behavior. 

@CaseyCarter, does this makes sense? (or more sense as before?).

> What they can't do is get a closed range that covers the entire range of integers without fear of overflow.

For unsigned integers they can only cover a sub-range of (half - 1) size. I think its worth it to assert on this to prevent undefined behavior. 

> And I disapprove of the inconsistent naming, FWIW.

I've renamed `inclusive_indices` to `closed_indices` "for consistency", but I think that `inclusive` is a better name, and we are deprecating `closed_ints` anyways, so if we remove that at some point, we will be left with `closed_indices` only. I'd rather not make a breaking change there again to rename it to `inclusive_indices` if we happen to like that better. Idk, what do you guys like more? Are there better names?

File: include/range/v3/view/indices.hpp (hunk 16 above)
Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: IIRC this was actually a clang bug right? This shouldn't be allowed. I'll fix it.

Comment (gnzlbg) [inline include/range/v3/view/indices.hpp]: Fixed and rebased.

**Comment by gnzlbg** [comment]
It seems that these are not true on MacOS :'( whyyyyyy!

```shell
/Users/travis/build/gnzlbg/range-v3/test/view/iota.cpp:122:5: error: static_assert failed ""
    static_assert(Same<std::intmax_t, range_difference_type_t<decltype(uints)>>(), "");
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/travis/build/gnzlbg/range-v3/test/view/iota.cpp:123:5: error: static_assert failed ""
    static_assert(Same<std::uintmax_t, range_size_type_t<decltype(uints)>>(), "");
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/travis/build/gnzlbg/range-v3/test/view/iota.cpp:129:5: error: static_assert failed ""
    static_assert(Same<std::intmax_t, range_difference_type_t<decltype(ints)>>(), "");
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/travis/build/gnzlbg/range-v3/test/view/iota.cpp:130:5: error: static_assert failed ""
    static_assert(Same<std::uintmax_t, range_size_type_t<decltype(ints)>>(), "");
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 errors generated.
```

**Comment by gnzlbg** [comment]
So, on MacOSX:

- `std::uint64_t == unsigned long long`
- `std::uintmax_t == unsigned long`
- `std::size_t == unsigned long`

and 

- `std::int64_t == long long`
- `std::intmax_t ==  long`
- `std::ptrdiff_t ==  long`

For `view::closed_ints`, however:

- `range_size_type_t<decltype(ints)> == unsigned long long`
- `range_difference_type_t<decltype(ints)> == long long`

because the difference type is based on `int_fastXX_t`, which on MacOSX is:

- `std::int_fast32_t == int`.
- `std::int_fast64_t == long long`. 

So... from `int_fast64_t` one gets `long long` and `unsigned long long` as difference and size types for `view::iota`, but `std::intmax_t` and `std::uintmax_t` are `long` and `unsigned long`. 

I think that the right thing to do is to replace `std::intmax_t` and `std::uintmax_t` with `std::int_fastXX_t` since that is actually what `view::iota` is doing.

**Comment by CaseyCarter** [comment]
> So, on MacOSX:
> ```c++
> std::uint64_t == unsigned long long
> std::uintmax_t == unsigned long
> std::size_t == unsigned long
> ```

It is insane - albeit conforming - that `intmax_t` has lower integer conversion rank than `long long`, even if `long` and `long long` do have the same object representation. I agree with your solution.

**Comment by ericniebler** [comment]
This one looks ready to go. I'll merge soon unless I hear shouts.

**Comment by gnzlbg** [comment]
@ericniebler there is a travis failure in the boost build bot, but i think this has always been the case (it surfaces now because clang 5.0 was broken for a week). For some reason Boost is failing to link, and I've debugged enough to know that it is not modules related (i can reproduce it without modules enabled on that build bot). I've fixed some parts of it in the cmake refactor PR, and plan to fully fix it there. That PR should deliver a green build on everything.
